### PR TITLE
Initial working implementation for security_seed (Issue #12)

### PR DESCRIPTION
**Pull request by bhass1**
Added new security_seed functionality to UDS module. This function asks the UDS server for a Security Access seed repeatedly and records every seed received.
It can be used to analyze the randomness of the UDS server's Security Access seed.
Supported options include different "Security Access levels", different "Diagnostic Sessions", and performing different types of "ECU Resets" between seed requests.
Results are printed to stdout.

image

image

### CODE REVIEW

**Comment by bhass1** [comment]
Fixes Issue #12

**Comment by kasperkarlsson** [comment]
Thank you very much for this contribution!

There are some issues which need to be resolved before this can be merged.

**General:**
- PEP 8 compliance; see e.g. https://pep8.org/ or https://www.python.org/dev/peps/pep-0008/ - some IDEs (e.g. PyCharm Community) even have built-in support for checking this.

**ISO-14229-1:**
It would be preferable if changes here were kept to a minimum.
- I do not see the need for the `is_valid_session` function, which just checks a function argument in the UDS module. Maybe it could be skipped altogether? 
- The comment change to `after CAN-TP layer has been removed` is redundant, as ISO-14229-1 works on another layer than the transport protocol.
- The `diagnostic_session_control` function looks great - keep as is! :+1:
- The default value `None` introduced for `security_access_request_seed` argument `data_record` is a correction, seeing as this is an optional parameter according to the specification. Great find! :+1: 

**UDS**
- The retry logic in `__security_seed_wrapper` feels quite arbitrary. Since this is in a loop, just remove it and make sure to verify that a response was received before handling it (otherwise it will be retried in the next iteration anyway). A single retry followed by forced handling may end up failing anyway.
- Hard-coded sleep time after `HARD_RESET`: timing may vary depending on target ECU and reset type. I would suggest taking the sleep duration as an optional command line argument to `cc.py uds security_seed` instead and apply it regardless of reset type, with some sensible default value (e.g. `0.1` seconds).
- Typo: *"secuirty"*
- The `print_negative_response` help function showing the meaning of an NFC is great :+1:

I am looking forward to accepting this PR! If you would like assistance to take care of the issues mentioned above, just let me know :smiley:

**Comment by bhass1** [comment]
Sorry for the force push, I messed up my rebase earlier! :weary: 

I haven't been able to test with a real CAN dongle yet, but I did some sanity checks and ran it through flake8... I will try to test with my two vehicles in the next week. :car: :car:

Thanks for the detailed feedback before!! It should all be addressed with the latest commits. Please let me know if you see any issues or want more changes.

**Comment by kasperkarlsson** [comment]
Great! I will review these new commits shortly :smiley:

**Comment by bhass1** [comment]
Tested on my vehicles. It works as intended, but I have to admit the user interface for the "sad" path is not the greatest. I'd like to make it more apparent for users when seeds _aren't_ being collected. It'd also be nice to build in a flag for fuzzing various security access levels until valid seeds start popping out...

Nonetheless I was still able to use it to find some weak implementations in my car tonight, so it works.

File: tool/modules/uds.py [hunk 1]
```diff
@@ -184,32 +198,47 @@ def is_valid_response(message):
...
-                            # Give some extra time for verification, in case of slow responses
-                            verification_end_time = time.time() + delay + VERIFICATION_EXTRA_DELAY
+                                print("    Resending 0x{0:0x}... ".format(
+                                  verify_arb_id), end=" ")
+                            tp.transmit(session_control_frames,
+                                        verify_arb_id, None)
+                            # Give some extra time for verification, in
+                            # case of slow responses
+                            verification_end_time = time.time()
+                            + delay + VERIFICATION_EXTRA_DELAY
```
Comment (kasperkarlsson) [inline tool/modules/uds.py]: This statement is now broken, as verification_end_time is simply set to `time.time()`. The delays on the next row are no longer added to this value.

File: tool/modules/uds.py [hunk 2]
```diff
@@ -619,94 +705,140 @@ def send_key(arb_id_request, arb_id_response, level, key, timeout):
...
-  cc.py uds discovery -autoblacklist 10
-  cc.py uds services 0x733 0x633
-  cc.py uds ecu_reset 1 0x733 0x633
-  cc.py uds testerpresent 0x733
-  cc.py uds security_seed 0x3 0x1 0x733 0x633 -r 1""")
+    parser = argparse.ArgumentParser(
+                prog="cc.py uds",
+                formatter_class=argparse.RawDescriptionHelpFormatter,
+                description="Universal Diagnostic Services module for"
+                "CaringCaribou",
```
Comment (kasperkarlsson) [inline tool/modules/uds.py]: Whitespace between `for` and `CaringCaribou` is missing

File: tool/modules/uds.py [hunk 3]
```diff
@@ -619,94 +705,140 @@ def send_key(arb_id_request, arb_id_response, level, key, timeout):
...
+                epilog="""Example usage:
+                cc.py uds discovery
+                cc.py uds discovery -blacklist 0x123 0x456
+                cc.py uds discovery -autoblacklist 10
+                cc.py uds services 0x733 0x633
+                cc.py uds ecu_reset 1 0x733 0x633
+                cc.py uds testerpresent 0x733
+                cc.py uds security_seed 0x3 0x1 0x733 0x633 -r 1""")
```
Comment (kasperkarlsson) [inline tool/modules/uds.py]: A lot of leading whitespace has been added to every single line of this multiline string, thus messing up the terminal output.

File: tool/lib/iso14229_1.py [hunk 4]
```diff
@@ -296,7 +292,9 @@ def is_positive_response(response):
         :return: False if response is a NEGATIVE_RESPONSE,
                  True otherwise
         """
-        if response is not None and len(response) > 0 and response[0] != Constants.NR_SI:
+        if(response is not None and
```
Comment (kasperkarlsson) [inline tool/lib/iso14229_1.py]: Missing whitespace between `if` and `(`

File: tool/lib/iso14229_1.py [hunk 5]
```diff
@@ -427,17 +434,21 @@ def input_output_control_by_identifier(self, identifier, data):
...
+        'identifier'
 
         :param identifier: DDDID to set
         :param sub_function: Sub function
         :param sub_function_arg: Sub function arguments
         :return: Response data if successful,
                  None otherwise
         """
-        if identifier is None or sub_function is None or sub_function_arg is None:
+        if(identifier is None or
```
Comment (kasperkarlsson) [inline tool/lib/iso14229_1.py]: Missing whitespace between `if` and `(`

File: tool/lib/iso14229_1.py [hunk 6]
```diff
@@ -542,7 +556,9 @@ def read_data_by_periodic_identifier(self, transmission_mode, identifier):
         :return: Response data if successful,
                  None otherwise
         """
-        if transmission_mode is None or identifier is None or len(identifier) == 0:
+        if(transmission_mode is None or
```
Comment (kasperkarlsson) [inline tool/lib/iso14229_1.py]: Missing whitespace between `if` and `(`

File: tool/modules/uds.py [hunk 7]
```diff
@@ -309,11 +346,14 @@ def service_discovery(arb_id_request, arb_id_response, timeout, min_id=BYTE_MIN,
...
                 # Parse response
                 if len(msg.data) > 3:
-                    # Since service ID is included in the response, mapping is correct even if response is delayed
+                    # Since service ID is included in the response, mapping
+                    # is correct even if response is delayed
                     service_id = msg.data[2]
                     status = msg.data[3]
-                    if status != NegativeResponseCodes.SERVICE_NOT_SUPPORTED:
-                        # Any other response than "service not supported" counts
+                    if(status !=
```
Comment (kasperkarlsson) [inline tool/modules/uds.py]: Missing whitespace between `if` and `(`

File: tool/modules/uds.py [hunk 8]
```diff
@@ -444,31 +501,39 @@ def __ecu_reset_wrapper(args):
...
         elif Iso14229_1.is_positive_response(response):
             # Positive response handling
             response_service_id = response[0]
             subfunction = response[1]
-            expected_response_id = Iso14229_1.get_service_response_id(Services.EcuReset.service_id)
-            if response_service_id == expected_response_id and subfunction == reset_type:
+            expected_response_id = \
+                Iso14229_1.get_service_response_id(
+                                               Services.EcuReset.service_id)
+            if(response_service_id == expected_response_id
```
Comment (kasperkarlsson) [inline tool/modules/uds.py]: Missing whitespace between `if` and `(`

File: tool/modules/uds.py [hunk 9]
```diff
@@ -534,27 +602,37 @@ def __security_seed_wrapper(args):
         for seed in seed_list:
             print(seed)
 
+
 def extended_session(arb_id_request, arb_id_response, session_type):
     # Sanity checks
-    if not Services.DiagnosticSessionControl.DiagnosticSessionType().is_valid_session(session_type):
-        raise ValueError("Invalid extended session type: 0x{0:02x}".format(session_type))
+    if(not Services.DiagnosticSessionControl.DiagnosticSessionType()
```
Comment (kasperkarlsson) [inline tool/modules/uds.py]: Missing whitespace between `if` and `(`

File: tool/modules/uds.py [hunk 10]
```diff
@@ -564,12 +642,15 @@ def request_seed(arb_id_request, arb_id_response, level, data_record, timeout):
     :rtype [int] or None
     """
     # Sanity checks
-    if not Services.SecurityAccess.RequestSeedOrSendKey().is_valid_request_seed_level(level):
+    if(not Services.SecurityAccess.RequestSeedOrSendKey()
```
Comment (kasperkarlsson) [inline tool/modules/uds.py]: Missing whitespace between `if` and `(`

File: tool/modules/uds.py [hunk 11]
```diff
@@ -600,12 +683,15 @@ def send_key(arb_id_request, arb_id_response, level, key, timeout):
     :rtype [int] or None
     """
     # Sanity checks
-    if not Services.SecurityAccess.RequestSeedOrSendKey().is_valid_send_key_level(level):
+    if(not Services.SecurityAccess.RequestSeedOrSendKey()
```
Comment (kasperkarlsson) [inline tool/modules/uds.py]: Missing whitespace between `if` and `(`

File: tool/modules/uds.py [hunk 12]
```diff
@@ -444,31 +501,39 @@ def __ecu_reset_wrapper(args):
             print("Received empty response")
         elif response_length == 1:
             # Invalid response length
-            print("Received response [{0:02x}] (1 byte), expected at least 2 bytes".format(response[0], len(response)))
+            print("Received response [{0:02x}] (1 byte), expected at least "
+                  " 2 bytes".format(response[0], len(response)))
```
Comment (kasperkarlsson) [inline tool/modules/uds.py]: Double whitespace between `at least` and `2 bytes`

File: tool/modules/uds.py [hunk 13]
```diff
@@ -329,41 +369,50 @@ def __service_discovery_wrapper(args):
...
 
-    # Calculate end timestamp if the TesterPresent should automatically stop after a given duration
+    # Calculate end timestamp if the TesterPresent should automatically
+    # stop after a given duration
     auto_stop = duration is not None
     end_time = None
     if auto_stop:
-        end_time = datetime.datetime.now() + datetime.timedelta(seconds=duration)
+        end_time = datetime.datetime.now()
+        + datetime.timedelta(seconds=duration)
```
Comment (kasperkarlsson) [inline tool/modules/uds.py]: Broken statement - the delta is no longer added...

**Comment by kasperkarlsson** [comment]
Thank you very much for your contribution, @bhass1! :smiley_cat:

**Pull request merged**